### ISSUE TITLE
pointer swizzling-based memory usage reduction

### ISSUE THREAD

**Issue by spacejam** [spacejam/sled#1093]
memory is sometimes exhausted when inserting a lot of items:
https://github.com/spacejam/sled/issues/1036
https://github.com/spacejam/sled/issues/986

nearly all memory is consumed by empty paged-out pages. the red majority below is just empty pages allocated in `page_out`.

![image](https://user-images.githubusercontent.com/1505488/83737975-01730400-a654-11ea-91b3-2b5425fa35c8.png)

systems like [umbra](http://cidrdb.org/cidr2020/papers/p29-neumann-cidr20.pdf) and llama use swizzling to just use a single pointer with some tag bits indicating whether the address being pointed to is in-memory or on-disk.

we have 4 states:
* paged-out with a single location
* paged-out with multiple locations
* paged-in with a single location
* paged-in with multiple locations

we can use 2 tag bits on an atomic pointer to point to one of 4 structs that contain the minimal required amount of information for using the page

1. paged-out single location can just be a u6(2)
2. paged-out multiple locations can point to a Vec<locations>
3. paged-in single location needs:
  * the page data
  * the logical timestamp (honestly this can be a u16 salt, does not need to be monotonic)
  * the location
4. paged-in multiple locations needs:
  * the page data
  * the logical timestamp
  * Vec<locations>

A location is used for:
1. pulling the page from disk. right now it's an enum that may point to a log coordinate or a blob file

We may have some parts of the codebase that depend on locations for LRU sizing (which is totally busted anyway) and file segment tenancy, but these dependencies should be broken if possible.

**Comment by spacejam** [comment]
after #1095, we're at this memory usage before employing swizzling

![massif_sled_after](https://user-images.githubusercontent.com/1505488/83875664-cc3fe200-a737-11ea-873a-9188f8d4b21d.png)

**Comment by ysimonson** [comment]
I'm seeing this on 0.34.6. In the baseline version of my project, memory use increases but it does so slowly. On an [experimental branch](https://github.com/indradb/indradb/pull/105), sled quickly exhausts memory. The differences between the two branches with respect to sled are:

- Instead of using multiple trees, only one is used.
- Batched writes are employed.

**Comment by calmofthestorm** [comment]
This has been a problem for me on and off on a few different projects, FWIW. I have a few systems with only 512 MB of total memory that I'd like to be able to use Sled on, but have found it challenging to avoid OOMs, even though I'm willing to accept very slow performance.

[Here is a simple example](https://gist.github.com/calmofthestorm/45ec06cc49c87cb2800bd6686e3618a4). `sled_oom <path> gen` will generate an ~11 GB database consisting of 64_000_000 small objects in a single tree. Running `sled_oom <path> oom` will then attempt to clear and then flush the tree. I have configured `cache_capacity` to use 4 MB (4_194_304). I'm actually fine with it going up to 128 MB or so, but am using a lower value to try to reduce the memory usage.

Both `gen` and `oom` use around 2.3 GB RSS and 3 GB VIRT. `gen` completes reasonably quickly (even without the parallelization), but `oom` is very slow (I gave up at 10 minutes). Slow is fine, but the memory usage is a problem. (I also find clear being so much more expensive than insertion strange, but can work around that with a generation count and parallelism for my particular use case).

It is possible that my use case is not a good fit for Sled, especially given I don't need it to be super fast, and went to a disk-backed database more out of memory constraints (though the persistence has other benefits). I could understand if this pattern is out of scope, and could probably write a simple custom data store that precisely controls memory, since slow is acceptable. But I did want to contribute this example in the hope that it may be useful should you decide to attempt to better support memory control and/or many small objects.

**Comment by calmofthestorm** [comment]
Having studied this more, I think I had simply chosen the wrong tool for the job. I rewrote it with LMDB and am very happy with the result -- memory usage is excellent thanks to mmap and performance is significantly faster. Sled is fantastic at what it does, but there may be better choices for your use case if your use case is mostly read only access to many small objects where only a single thread may write at a time is acceptable and you're willing to deal with a bit clunkier of an API and accept native code dependencies. The read only snapshots have also allowed me to make some other optimizations.

**Comment by uuhan** [comment]
I have encounted the same memory issue in my project. 

As time goes by, my sled database reaches 12GB size(with compression enabled). Every time I start my app, sled takes 8GB of my memory. The memory usage gose out of control with the more insertions and the memory consumpation will grow faster and faster (emm, like exponential function), finally it is killed by OOM.

My sled version is 0.34.6, and I have nearly 400 trees opened in my server app and insert 100 items per minute into each of them.

May be sled is not suit for my case. I'll try to use rocksdb as an alternative to see the memory consumpation.

**Comment by emnvn** [comment]
I m facing the memory issues with Sled, in my use case, my service read cameras frame then save to sled db,  for each camera, we have each database  per day. We insert 5 frames per second per camera, the result we found that memory is not released, it will consume more than 95% memory after 2 hours (my computer has 16 GB RAM). 

We are looking the solution for my issue.
